File: src/main/java/org/luxons/sevenwonders/game/data/CardDefinition.java
package org.luxons.sevenwonders.game.data;

import java.util.List;
import java.util.Map;

import org.luxons.sevenwonders.game.cards.Card;
import org.luxons.sevenwonders.game.cards.Color;
import org.luxons.sevenwonders.game.cards.Requirements;

public class CardDefinition {

    private String name;

    private Color color;

    private String effect;

    private Requirements requirements;

    private String chainParent;

    private List<String> chainChildren;

    private Map<Integer, Integer> countPerNbPlayer;

    public Card createCard() {
        return new Card(name, color, requirements, EffectParser.parse(effect));
    }

    private Requirements createRequirements() {
        Requirements req = new Requirements();

        return req;
    }

    public String getName() {
        return name;
    }

    public Color getColor() {
        return color;
    }

    public String getEffect() {
        return effect;
    }

    public Requirements getRequirements() {
        return requirements;
    }

    public String getChainParent() {
        return chainParent;
    }

    public List<String> getChainChildren() {
        return chainChildren;
    }

    public Map<Integer, Integer> getCountPerNbPlayer() {
        return countPerNbPlayer;
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/game/data/definitions/CardDefinition.kt
package org.luxons.sevenwonders.game.data.definitions

import org.luxons.sevenwonders.game.cards.Card
import org.luxons.sevenwonders.game.cards.CardBack
import org.luxons.sevenwonders.game.cards.Color
import org.luxons.sevenwonders.game.cards.Requirements

internal class CardDefinition(
    private val name: String,
    private val color: Color,
    private val requirements: Requirements?,
    private val effect: EffectsDefinition,
    private val chainParent: String?,
    private val chainChildren: List<String>?,
    private val image: String,
    private val countPerNbPlayer: Map<Int, Int>
) {
    fun create(back: CardBack, nbPlayers: Int): List<Card> = List(countPerNbPlayer[nbPlayers] ?: 0) { create(back) }

    fun create(back: CardBack): Card {
        val reqs = requirements ?: Requirements()
        val children = chainChildren ?: emptyList()
        return Card(name, color, reqs, effect.create(), chainParent, children, image, back)
    }
}
